### PR TITLE
Update testing.Clock to improve robustness

### PR DESCRIPTION
**Pull request by reedobrien**
Adds a WaitAdvance method which ensures that there is something in
clock.waiting before actually advancing the time. Addtionally cause
Advance to log a debug message if it Advances time when clock.waiting
is empty.

Refs: https://bugs.launchpad.net/juju/+bug/1607044

QA:
1) Run the unit tests and ensure they all pass.

### CODE REVIEW

File: clock.go [hunk 1]
```diff
@@ -102,9 +103,40 @@ func (clock *Clock) Advance(d time.Duration) {
 	clock.mu.Lock()
 	defer clock.mu.Unlock()
 	clock.now = clock.now.Add(d)
+	if len(clock.waiting) == 0 {
+		logger.Debugf("advancing a clock that has nothing waiting")
```
Comment (reedobrien) [inline clock.go]: Create a wiki entry about this and point this message at it.

Comment (reedobrien) [inline clock.go]: Create a wiki entry about this and point this message at it.

File: clock_test.go [hunk 2]
```diff
@@ -4,14 +4,18 @@
 package testing_test
 
 import (
-	gc "gopkg.in/check.v1"
 	"sync"
 	"time"
 
+	gc "gopkg.in/check.v1"
+
 	"github.com/juju/testing"
```
Comment (anastasiamac) [inline clock_test.go]: import block :D

File: clock.go [hunk 3]
```diff
@@ -102,9 +103,41 @@ func (clock *Clock) Advance(d time.Duration) {
...
+	}
+	pause := w / 10
+	for i := 0; i < 10; i++ {
+		if clock.hasNWaiters(n) {
+			clock.Advance(d)
+			return nil
+		}
+		time.Sleep(pause)
+	}
+	return errors.Errorf("no timers added after waiting %s", w.String())
```
Comment (babbageclunk) [inline clock.go]: This needs to be updated for the `n` waiters change. It's probably worth saying how many waiters there were. Of course, then you'll get the irritating situation where one shows up between the last check and the report: "waiting for 10 timers, only 10 seen".

File: clock_test.go (hunk 2 above)
Comment (reedobrien) [inline clock_test.go]: As discussed in person at the sprint, this conforms with how block are separated in juju/cmd and other libraries outside of juju/juju. Also, it is how goimports auto formats imports. :)

File: clock.go (hunk 3 above)
Comment (reedobrien) [inline clock.go]: updated error message

**Comment by reedobrien** [comment]
$$merge$$

**Pull request merged**